File: NapakalakiRuby/lib/cultist_player.rb
# To change this license header, choose License Headers in Project Properties.
# To change this template file, choose Tools | Templates
# and open the template in the editor.
require_relative 'player'
require_relative 'cultist'

module NapakalakiGame
  class CultistPlayer < Player
    @@totalCultistPlayers = 0
    
    def initialize(p,c)
      super(p.name)
      copyAll(p)
      @myCultistCard = c
      @@totalCultistPlayers += 1
    end
    
    def copyAll(p)
      @dead = p.dead
      @name = p.name
      @level = p.level
      @hiddenTreasures = p.hiddenTreasures
      @visibleTreasures = p.visibleTreasures
      @pendingBadConsequence = p.pendingBadConsequence
      @enemy = p.enemy
      @canISteal = p.canISteal
    end
    
    def self.getTotalCultistPlayers
      @@totalCultistPlayers
    end
    
    def getCombatLevel
      nivel_superior = super
      nivel_carta = @myCultistCard.getGainedLevels
      
      return (nivel_superior + 0.2*nivel_superior + nivel_carta*@@totalCultistPlayers).round
    end
    
    def canYouGiveMeATreasure
      return (!getVisibleTreasures.empty?);
    end
    
    def giveMeATreasure
      aleatorio = rand(@visibleTreasures.size)
      tesoro = @visibleTreasures.at(aleatorio)
      nuevo_tesoro = Treasure.new(tesoro.name, tesoro.bonus, tesoro.type)
      @visibleTreasures.delete(tesoro)
      
      return nuevo_tesoro
    end
    
    def shouldConvert
      return false
    end
    
    def getOponentLevel(m)
      return m.getCombatLevelAgainstCultistPlayer
    end
    
    def to_s
      return  "name: #{@name}
      level: #{@level}
      combatLevel: #{getCombatLevel}
      dead: #{@dead}
      canISteal: #{@canISteal}
      pendingBC #{@pendingBadConsequence}
      Soy un jugador cultist.
      Carta: #{@myCultistCard}"
    end
  end
end

File: NapakalakiJava/src/NapakalakiGame/CultistPlayer.java
/*
 * To change this license header, choose License Headers in Project Properties.
 * To change this template file, choose Tools | Templates
 * and open the template in the editor.
 */
package NapakalakiGame;

import java.util.Random;

/**
 *
 * @author danidiaz
 */
public class CultistPlayer extends Player
{
    private static int totalCultistPlayers = 0;
    private Cultist myCultistCard;
    
    /**
     * @brief Constructor
     */
    public CultistPlayer(Player p, Cultist c)
    {
        super(p);
        myCultistCard = c;
        totalCultistPlayers++;
    }
    
    /**
     * @brief Consultor de cuantos jugadores cultist hay.
     * @return 
     */
    public static int getTotalCultistPlayers()
    {
        return totalCultistPlayers;
    }
    
    /**
     * @brief Redefinicion del metodo getCombatLevel
     * @return 
     */
    @Override
    public int getCombatLevel()
    {
        int nivel_superior = super.getCombatLevel();
        int nivel_carta = myCultistCard.getGainedLevels();
        
        return (int) (nivel_superior + 0.2*nivel_superior + 
                nivel_carta*totalCultistPlayers);
    }
    
    /**
     * @brief Redefinición del método de la clase superior.
     *      Se hace lo mismo, pero con los tesoros visibles.
     * @return 
     */
    @Override
    protected boolean canYouGiveMeATreasure()
    {
        return (!getVisibleTreasures().isEmpty());
    }
    
    /**
     * @brief Redefinición del método de la clase superior.
     *      Se hace lo mismo, pero con los tesoros visibles.
     * @return 
     */
    @Override
    protected Treasure giveMeATreasure()
    {
        int aleatorio = new Random().nextInt(getVisibleTreasures().size());
        Treasure tesoro = getVisibleTreasures().get(aleatorio);
        Treasure nuevo_tesoro = new Treasure(tesoro.getName(),tesoro.getBonus(),tesoro.getType());
        getVisibleTreasures().remove(tesoro);

        return (nuevo_tesoro);
    }
    
    /**
     * @brief El jugador no puede convertirse en cultist porque ya lo es.
     * @return Siempre falso.
     */
    @Override
    protected boolean shouldConvert()
    {
        return false;
    }
    
    /**
     * @brief Devuelve el nivel del combate contra cultist player del monstruo
     * @param m Monstruo a pedir el nivel
     * @return Nivel del monstruo contra jugador cultist
     */
    @Override
    protected int getOponentLevel(Monster m)
    {
        return m.getCombatLevelAgainstCultistPlayer();
    }
    
    @Override
    public String toString()
    {
        return super.toString() + "\nSoy cultist player" +
                "\nCarta: " + myCultistCard.toString();
    }
}
